### PR TITLE
Calls to panic replaced with returning an error 

### PR DESCRIPTION
**Pull request by MarkSonghurst**

### CODE REVIEW

File: error.go [hunk 1]
```diff
@@ -55,7 +58,7 @@ func NewError(apiName string, handle interface{}) error {
 			break
 		}
 		if IsError(ret) {
-			panic(fmt.Errorf("SQLGetDiagRec failed: ret=%d", ret))
+			return fmt.Errorf("SQLGetDiagRec failed: ret=%d", ret)
```
Comment (alexbrainman) [inline error.go]: NewError function is called when `handle` parameter has outstanding error condition on it. The purpose of NewError is to retrieve that error and return it to the caller. But you are replacing that error with completely new error returned from SQLGetDiagRec. Why it is OK to silently replace the error? What about original error? Wouldn't caller need to know what the original error was?

**Comment by MarkSonghurst** [comment]
Thanks for the review. However, I'm struggling to understand your comment.
The original code would call panic with the value of ret and terminate the program (assuming there was no recover to catch the panic), where as my replacement does much the same thing, except it returns the value of ret in an error.
Looking at the failure, it would seem there is no real error to return if SQLGetDiagRec has failed to return without ret being assigned an error code. So the error within handle cannot be extracted anyway?

**Comment by MarkSonghurst** [comment]
I can put the apiName into the generated error, maybe that would be sufficient?
Something like (line 61)
return fmt.Errorf("SQLGetDiagRec failed: ret=%d (original error in %s)", ret, apiName)

**Comment by alexbrainman** [comment]
> The original code would call panic with the value of ret and terminate the program (assuming there was no recover to catch the panic), where as my replacement does much the same thing, except it returns the value of ret in an error.

If you are package user, what would you do with that error? Lets say you have SQL with syntax error, and you get error that says: "SQLGetDiagRec failed: ret=-2". What would your program do to handle this error?

Alex

**Comment by MarkSonghurst** [comment]
My daemon would print the message in my log at error level and raise an alert for an sysadmin to investigate. Then I would return an HTTP 500 series error message to the client at the other end of my REST API.

**Comment by alexbrainman** [comment]
> My daemon would print the message in my log at error level and raise an alert for an sysadmin to investigate. Then I would return an HTTP 500 series error message to the client at the other end of my REST API.

I don't see how any of these are different from panicking. You could setup your daemon to log stack trace of the crash, send this information to sysadmin to investigate, and restart the program. Why your approach is better?

Alex

**Comment by MarkSonghurst** [comment]
A panic terminates the program. At any one time my daemon could be processing many requests to many clients. All those requests would be killed by the crash.

**Comment by MarkSonghurst** [comment]
I'll maintain my own fork of this package. Thanks for your time.

**Comment by alexbrainman** [comment]
I am sorry, it took me awhile to decide. But if you prefer to keep your own fork, I am fine with that too.

Alex

**Comment by MarkSonghurst** [comment]
What did you decide? To merge, or not?

**Comment by alexbrainman** [comment]
I don't have time to discuss now, but I will tell you within one day. If you want to leave it open, I will play with it. Thank you.

Alex

**Comment by MarkSonghurst** [comment]
I've re-opened the PR, but you will get other unrelated commits I made to my master now.
Hopefully you can take just the commits you want easily.

**Comment by alexbrainman** [comment]
I tried running tests against your changes, but I get this:
```
# github.com/alexbrainman/odbc
.\errorhandling_test.go:29: cannot use nil as type api.SQLHSTMT in argument to b.Value
.\errorhandling_test.go:43: cannot convert nil to type api.SQLHANDLE
.\errorhandling_test.go:52: cannot use nil as type api.SQLHSTMT in argument to p.BindValue
FAIL    github.com/alexbrainman/odbc [build failed]
# github.com/alexbrainman/odbc
.\errorhandling_test.go:29: cannot use nil as type api.SQLHSTMT in argument to b.Value
.\errorhandling_test.go:43: cannot convert nil to type api.SQLHANDLE
.\errorhandling_test.go:52: cannot use nil as type api.SQLHSTMT in argument to p.BindValue
FAIL    github.com/alexbrainman/odbc [build failed]
# github.com/alexbrainman/odbc
.\errorhandling_test.go:29: cannot use nil as type api.SQLHSTMT in argument to b.Value
.\errorhandling_test.go:43: cannot convert nil to type api.SQLHANDLE
.\errorhandling_test.go:52: cannot use nil as type api.SQLHSTMT in argument to p.BindValue
FAIL    github.com/alexbrainman/odbc [build failed]
```
Also, please, remove everything unrelated to the subject of this PR from this.

Thank you.

Alex

**Comment by MarkSonghurst** [comment]
I cannot figure out a way to make my tests support both Linux and Windows.
It's due to the type differences in the api directory which Go objects to at compile time. 
This doesn't work:

```
	if runtime.GOOS == "windows" {
		v, err = b.Value(api.SQL_NULL_HSTMT, 0)
	} else {
		v, err = b.Value(nil, 0)
	}

```


I'm closing the pull request; i'll maintain my own fork from here on in, which I guess will now always be Linux specific.
Thanks for your time anyway.

**Comment by alexbrainman** [comment]
You could drop new tests altogether. But keeping private fork is fine with me too.

Alex

**Comment by MarkSonghurst** [comment]
Hi @alexbrainman 
If I reopen this PR and do your suggestions made here:

> You could drop new tests altogether.
> Also, please, remove everything unrelated to the subject of this PR from this.

Will you reconsider this PR? I saw some other people were interested it in #76 so it seems fair for me to share it upstream to you.

**Comment by alexbrainman** [comment]
> Will you reconsider this PR?

I am happy to review it again, but I won't have free time until maybe weekend. And, please, remove everything unrelated to removing panics.

Thank you.

Alex

**Pull request closed without merging**